--- source/creator/CMakeLists.txt   2012-08-29 15:27:23.861123455 +0400
+++ source/creator/CMakeLists.txt   2012-08-29 15:25:51.541125000 +0400
@@ -841,6 +841,8 @@
 		bf_editor_mask
 		bf_editor_io
 
+		vb_exporter
+
 		bf_render
 		bf_intern_opennl
 		bf_python

--- source/blender/SConscript	2011-11-10 04:15:47.023294417 +0400
+++ source/blender/SConscript	2011-11-26 18:47:33.622125572 +0400
@@ -10,6 +10,7 @@
             'blenpluginapi/SConscript',
             'gpu/SConscript',
             'editors/SConscript',
+            'exporter/SConscript',
             'imbuf/SConscript',
             'makesdna/SConscript',
             'render/SConscript',

--- source/blender/CMakeLists.txt	2011-11-20 00:30:15.210801416 +0400
+++ source/blender/CMakeLists.txt	2011-11-26 18:54:30.432112315 +0400
@@ -106,6 +106,7 @@
 add_subdirectory(modifiers)
 add_subdirectory(makesdna)
 add_subdirectory(makesrna)
+add_subdirectory(exporter)
 
 if(WITH_COMPOSITOR)
 	add_subdirectory(opencl)  # later on this may be used more generally

--- source/blender/editors/space_api/CMakeLists.txt	2011-10-25 11:38:16.371086801 +0400
+++ source/blender/editors/space_api/CMakeLists.txt	2011-11-26 18:52:03.095117001 +0400
@@ -29,6 +29,7 @@
 	../../makesrna
 	../../windowmanager
 	../../../../intern/guardedalloc
+	../../exporter
 )

 set(INC_SYS

--- source/blender/editors/space_api/spacetypes.c   2012-08-29 15:35:17.173114053 +0400
+++ source/blender/editors/space_api/spacetypes.c   2012-09-06 15:07:51.672606940 +0400
@@ -65,6 +65,8 @@
 #include "ED_mask.h"
 #include "ED_sequencer.h"
 
+#include "ED_exporter.h"
+
 #include "io_ops.h"
 
 /* only call once on startup, storage is global in BKE kernel listbase */
@@ -98,6 +100,7 @@
 //	...
 	
 	/* register operator types for screen and all spaces */
+	ED_operatortypes_exporter();
 	ED_operatortypes_screen();
 	ED_operatortypes_anim();
 	ED_operatortypes_animchannels();

--- source/blender/editors/space_api/SConscript	2013-05-26 17:22:25.962599458 +0400
+++ source/blender/editors/space_api/SConscript	2013-05-26 17:27:31.125898040 +0400
@@ -42,6 +42,7 @@
     '../../makesrna',
     '../../python',
     '../../windowmanager',
+    '../../exporter'
     ]
 incs = ' '.join(incs)
 

--- source/blender/makesrna/intern/rna_texture.c	2011-11-26 18:38:44.088142414 +0400
+++ source/blender/makesrna/intern/rna_texture.c	2011-11-26 18:59:28.344102839 +0400
@@ -56,6 +56,7 @@

 EnumPropertyItem texture_type_items[] = {
 	{0, "NONE", 0, "None", ""},
+	{TEX_VRAY, "VRAY", ICON_TEXTURE, "V-Ray", ""},
 	{TEX_BLEND, "BLEND", ICON_TEXTURE, "Blend", "Procedural - create a ramp texture"},
 	{TEX_CLOUDS, "CLOUDS", ICON_TEXTURE, "Clouds", "Procedural - create a cloud-like fractal noise texture"},
 	{TEX_DISTNOISE, "DISTORTED_NOISE", ICON_TEXTURE,

--- source/blender/makesdna/DNA_texture_types.h	2011-11-20 00:30:04.470801629 +0400
+++ source/blender/makesdna/DNA_texture_types.h	2011-11-26 19:00:38.076100621 +0400
@@ -290,6 +290,7 @@
 /* **************** TEX ********************* */

 /* type */
+#define TEX_VRAY		200
 #define TEX_CLOUDS		1
 #define TEX_WOOD		2
 #define TEX_MARBLE		3

--- source/blender/python/intern/bpy_rna.c	2011-11-26 18:38:41.178142507 +0400
+++ source/blender/python/intern/bpy_rna.c	2011-11-26 19:04:03.191094097 +0400
@@ -81,7 +81,7 @@
 #include "BLF_translation.h"
 #endif

-#define USE_PEDANTIC_WRITE
+//#define USE_PEDANTIC_WRITE
 #define USE_MATHUTILS
 #define USE_STRING_COERCE

--- source/blender/editors/include/UI_icons.h	2013-03-21 21:26:15.480537000 +0400
+++ source/blender/editors/include/UI_icons.h	2013-05-27 19:19:23.961581510 +0400
@@ -279,9 +279,11 @@
 DEF_ICON(RNA)
 DEF_ICON(RNA_ADD)
 
+	/* V-Ray Icons */
+DEF_ICON(VRAY_LOGO)
+
 	/* available */
 #ifndef DEF_ICON_BLANK_SKIP
-	DEF_ICON(BLANK092)
 	DEF_ICON(BLANK093)
 	DEF_ICON(BLANK094)
 	DEF_ICON(BLANK095)

