--- source/blender/SConscript 2011-01-06 14:52:13.000000000 +0300
+++ source/blender/SConscript  2011-01-06 15:13:12.474295043 +0300
@@ -9,6 +9,7 @@
             'blenpluginapi/SConscript',
             'gpu/SConscript',
             'editors/SConscript',
+            'exporter/SConscript',
             'imbuf/SConscript',
             'makesdna/SConscript',
             'makesrna/SConscript',

--- source/blender/editors/space_api/spacetypes.c	2011-06-05 16:57:09.124312000 +0400
+++ source/blender/editors/space_api/spacetypes.c	2011-06-08 10:51:47.875149901 +0400
@@ -64,6 +64,8 @@
 #include "ED_mball.h"
 #include "ED_logic.h"
 
+#include "exporter_ops.h"
+
 /* only call once on startup, storage is global in BKE kernel listbase */
 void ED_spacetypes_init(void)
 {
@@ -112,6 +114,7 @@
 	ED_operatortypes_sound();
 	ED_operatortypes_render();
 	ED_operatortypes_logic();
+	ED_operatortypes_exporter(); // V-Ray/Blender
 	
 	UI_view2d_operatortypes();
 	UI_buttons_operatortypes();
 	
--- source/blender/editors/space_api/SConscript	2010-07-19 17:09:38.327627551 +0400
+++ source/blender/editors/space_api/SConscript	2010-07-19 17:20:02.931373123 +0400
@@ -4,6 +4,7 @@
 sources = env.Glob('*.c')
 
 incs = '../include ../../blenlib ../../blenkernel ../../blenloader ../../makesdna'
+incs += ' ../../exporter'
 incs += ' ../../windowmanager ../../python ../../makesrna'
 incs += ' #/intern/guardedalloc #/extern/glew/include'

--- source/blender/makesrna/intern/rna_texture.c	2011-03-02 02:40:21.756845001 +0300
+++ source/blender/makesrna/intern/rna_texture.c	2011-04-04 01:51:36.334316052 +0400
@@ -71,6 +71,7 @@ EnumPropertyItem texture_type_items[] = 
	{TEX_VORONOI, "VORONOI", ICON_TEXTURE, "Voronoi", "Procedural - Creates cell-like patterns based on Worley noise"},
	{TEX_VOXELDATA, "VOXEL_DATA", ICON_TEXTURE, "Voxel Data", "Creates a 3d texture based on volumetric data"},
	{TEX_WOOD, "WOOD", ICON_TEXTURE, "Wood", "Procedural - Wave generated bands or rings, with optional noise"},
+	{TEX_VRAY, "VRAY", ICON_TEXTURE, "V-Ray", ""},
 	{0, NULL, 0, NULL, NULL}};
 
 #ifdef RNA_RUNTIME

--- source/blender/editors/space_api/CMakeLists.txt	2011-01-06 14:51:57.000000000 +0300
+++ source/blender/editors/space_api/CMakeLists.txt	2011-01-14 18:03:57.457981234 +0300
@@ -24,6 +24,7 @@ set(INC
 	../../blenkernel
 	../../blenloader
 	../../blenlib
+	../../exporter
 	../../makesdna
 	../../makesrna
 	../../windowmanager

--- source/blender/CMakeLists.txt	2011-01-06 14:52:13.000000000 +0300
+++ source/blender/CMakeLists.txt	2011-01-14 17:53:03.215981234 +0300
@@ -100,6 +100,7 @@
 add_subdirectory(modifiers)
 add_subdirectory(makesdna)
 add_subdirectory(makesrna)
+add_subdirectory(exporter)
 
 if(WITH_IMAGE_OPENEXR)
 	add_subdirectory(imbuf/intern/openexr)

--- source/creator/CMakeLists.txt	2011-01-06 14:48:46.000000000 +0300
+++ source/creator/CMakeLists.txt	2011-01-14 19:05:58.851581234 +0300
@@ -731,6 +754,7 @@
		bf_editor_space_time
		bf_editor_space_userpref
		bf_editor_space_view3d
+		vb_mesh_exporter
 
		bf_editor_text
		bf_editor_transform

--- source/blender/makesdna/DNA_texture_types.h	2011-01-07 22:42:51.812373001 +0300
+++ source/blender/makesdna/DNA_texture_types.h	2011-01-21 18:42:30.430959002 +0300
@@ -304,6 +304,7 @@
 #define TEX_DISTNOISE	13
 #define TEX_POINTDENSITY	14
 #define TEX_VOXELDATA		15
+#define TEX_VRAY		200
 
 /* musgrave stype */
 #define TEX_MFRACTAL		0

--- source/blender/python/intern/bpy_rna.c	2011-06-07 20:08:49.353294000 +0400
+++ source/blender/python/intern/bpy_rna.c	2011-06-08 10:49:16.776949902 +0400
@@ -72,7 +72,7 @@
 #include "../generic/IDProp.h" /* for IDprop lookups */
 #include "../generic/py_capi_utils.h"
 
-#define USE_PEDANTIC_WRITE
+/*#define USE_PEDANTIC_WRITE*/
 #define USE_MATHUTILS
 #define USE_STRING_COERCE
