--- source/blender/SConscript	2010-07-19 17:09:57.809133217 +0400
+++ SConscript	2010-07-19 17:30:30.617622329 +0400
@@ -16,6 +16,7 @@
             'readblenfile/SConscript',
             'render/SConscript',
             'nodes/SConscript',
+            'exporter/SConscript',
             'modifiers/SConscript',
             'ikplugin/SConscript',
             'windowmanager/SConscript',
--- source/blender/editors/space_api/spacetypes.c	2010-07-19 17:09:38.327627551 +0400
+++ space_api/spacetypes.c	2010-07-19 17:21:13.397622897 +0400
@@ -57,6 +57,8 @@
 #include "ED_mball.h"
 #include "ED_logic.h"
 
+#include "exporter_ops.h"
+
 /* only call once on startup, storage is global in BKE kernel listbase */
 void ED_spacetypes_init(void)
 {
@@ -102,7 +104,10 @@
 	ED_operatortypes_sound();
 	ED_operatortypes_render();
 	ED_operatortypes_logic();
-	
+
+    // V-Ray/Blender exporter operator
+	ED_operatortypes_exporter();
+
 	UI_view2d_operatortypes();
 	UI_buttons_operatortypes();
 	
--- source/blender/editors/space_api/SConscript	2010-07-19 17:09:38.327627551 +0400
+++ space_api/SConscript	2010-07-19 17:20:02.931373123 +0400
@@ -4,6 +4,7 @@
 sources = env.Glob('*.c')
 
 incs = '../include ../../blenlib ../../blenkernel ../../makesdna'
+incs += ' ../../exporter'
 incs += ' ../../windowmanager ../../python ../../makesrna'
 incs += ' #/intern/guardedalloc #/extern/glew/include'
 
--- source/blender/editors/render/render_internal.c	2010-08-13 04:48:14.330008395 +0400
+++ render_internal.c	2010-08-13 04:49:57.960078889 +0400
@@ -654,7 +654,7 @@
 	// store spare
 
 	/* ensure at least 1 area shows result */
-	screen_set_image_output(C, event->x, event->y);
+	/* screen_set_image_output(C, event->x, event->y); */
 
 	/* single layer re-render */
 	if(RNA_property_is_set(op->ptr, "layer")) {

--- source/blender/makesrna/intern/rna_texture.c	2010-09-01 14:02:55.819168558 +0400
+++ rna_texture.c	2010-09-01 14:04:33.795918790 +0400
@@ -66,6 +66,7 @@
 	{TEX_VORONOI, "VORONOI", ICON_TEXTURE, "Voronoi", ""},
 	{TEX_VOXELDATA, "VOXEL_DATA", ICON_TEXTURE, "Voxel Data", ""},
 	{TEX_WOOD, "WOOD", ICON_TEXTURE, "Wood", ""},
+	{200, "VRAY", ICON_TEXTURE, "V-Ray", ""},
 	{0, NULL, 0, NULL, NULL}};
 
 #ifdef RNA_RUNTIME
 
--- source/blender/python/intern/bpy_rna.c	2010-12-17 05:44:53.302238004 +0300
+++ bpy_rna.c	2010-12-17 19:59:02.645938003 +0300
@@ -53,7 +53,7 @@
 #include "../generic/IDProp.h" /* for IDprop lookups */
 #include "../generic/py_capi_utils.h"
 
-#define USE_PEDANTIC_WRITE
+//#define USE_PEDANTIC_WRITE
 #define USE_MATHUTILS
 #define USE_STRING_COERCE
 
