--- source/blender/SConscript 2011-01-06 14:52:13.000000000 +0300
+++ source/blender/SConscript  2011-01-06 15:13:12.474295043 +0300
@@ -9,6 +9,7 @@
             'blenpluginapi/SConscript',
             'gpu/SConscript',
             'editors/SConscript',
+            'exporter/SConscript',
             'imbuf/SConscript',
             'makesdna/SConscript',
             'makesrna/SConscript',

--- source/blender/editors/space_api/spacetypes.c	2010-07-19 17:09:38.327627551 +0400
+++ source/blender/editors/space_api/spacetypes.c	2010-07-19 17:21:13.397622897 +0400
@@ -59,6 +59,8 @@
 #include "ED_mball.h"
 #include "ED_logic.h"
 
+#include "exporter_ops.h"
+
 /* only call once on startup, storage is global in BKE kernel listbase */
 void ED_spacetypes_init(void)
 {
@@ -104,7 +106,10 @@
 	ED_operatortypes_sound();
 	ED_operatortypes_render();
 	ED_operatortypes_logic();
-	
+
+	// V-Ray/Blender mesh export operator
+	ED_operatortypes_exporter();
+
 	UI_view2d_operatortypes();
 	UI_buttons_operatortypes();
 	
--- source/blender/editors/space_api/SConscript	2010-07-19 17:09:38.327627551 +0400
+++ source/blender/editors/space_api/SConscript	2010-07-19 17:20:02.931373123 +0400
@@ -4,6 +4,7 @@
 sources = env.Glob('*.c')
 
 incs = '../include ../../blenlib ../../blenkernel ../../makesdna'
+incs += ' ../../exporter'
 incs += ' ../../windowmanager ../../python ../../makesrna'
 incs += ' #/intern/guardedalloc #/extern/glew/include'

--- source/blender/python/intern/bpy_rna.c	2010-12-17 05:44:53.302238004 +0300
+++ source/blender/python/intern/bpy_rna.c	2010-12-17 19:59:02.645938003 +0300
@@ -54,7 +54,7 @@
 #include "../generic/IDProp.h" /* for IDprop lookups */
 #include "../generic/py_capi_utils.h"
 
-#define USE_PEDANTIC_WRITE
+//#define USE_PEDANTIC_WRITE
 #define USE_MATHUTILS
 #define USE_STRING_COERCE

--- source/blender/editors/render/render_internal.c   2011-01-06 14:51:55.000000000 +0300
+++ source/blender/editors/render/render_internal.c   2011-01-06 16:08:02.447495042 +0300
@@ -699,7 +699,7 @@ static int screen_render_invoke(bContext
 	// store spare
 
 	/* ensure at least 1 area shows result */
-	screen_set_image_output(C, event->x, event->y);
+	/* screen_set_image_output(C, event->x, event->y); */
 
 	/* single layer re-render */
 	if(RNA_property_is_set(op->ptr, "layer")) {

--- source/blender/makesrna/intern/rna_texture.c	2011-01-06 14:51:10.000000000 +0300
+++ source/blender/makesrna/intern/rna_texture.c	2011-01-06 15:10:08.536095042 +0300
@@ -51,21 +51,22 @@ EnumPropertyItem texture_filter_items[] 
 
 EnumPropertyItem texture_type_items[] = {
 	{0, "NONE", 0, "None", ""},
-	{TEX_BLEND, "BLEND", ICON_TEXTURE, "Blend", ""},
-	{TEX_CLOUDS, "CLOUDS", ICON_TEXTURE, "Clouds", ""},
-	{TEX_DISTNOISE, "DISTORTED_NOISE", ICON_TEXTURE, "Distorted Noise", ""},
-	{TEX_ENVMAP, "ENVIRONMENT_MAP", ICON_IMAGE_DATA, "Environment Map", ""},
+	/* {TEX_BLEND, "BLEND", ICON_TEXTURE, "Blend", ""}, */
+	/* {TEX_CLOUDS, "CLOUDS", ICON_TEXTURE, "Clouds", ""}, */
+	/* {TEX_DISTNOISE, "DISTORTED_NOISE", ICON_TEXTURE, "Distorted Noise", ""}, */
+	/* {TEX_ENVMAP, "ENVIRONMENT_MAP", ICON_IMAGE_DATA, "Environment Map", ""}, */
 	{TEX_IMAGE, "IMAGE", ICON_IMAGE_DATA, "Image or Movie", ""},
-	{TEX_MAGIC, "MAGIC", ICON_TEXTURE, "Magic", ""},
-	{TEX_MARBLE, "MARBLE", ICON_TEXTURE, "Marble", ""},
-	{TEX_MUSGRAVE, "MUSGRAVE", ICON_TEXTURE, "Musgrave", ""},
-	{TEX_NOISE, "NOISE", ICON_TEXTURE, "Noise", ""},
+	/* {TEX_MAGIC, "MAGIC", ICON_TEXTURE, "Magic", ""}, */
+	/* {TEX_MARBLE, "MARBLE", ICON_TEXTURE, "Marble", ""}, */
+	/* {TEX_MUSGRAVE, "MUSGRAVE", ICON_TEXTURE, "Musgrave", ""}, */
+	/* {TEX_NOISE, "NOISE", ICON_TEXTURE, "Noise", ""}, */
 	//{TEX_PLUGIN, "PLUGIN", ICON_PLUGIN, "Plugin", ""}, /* Nothing yet */
-	{TEX_POINTDENSITY, "POINT_DENSITY", ICON_TEXTURE, "Point Density", ""},
-	{TEX_STUCCI, "STUCCI", ICON_TEXTURE, "Stucci", ""},
-	{TEX_VORONOI, "VORONOI", ICON_TEXTURE, "Voronoi", ""},
-	{TEX_VOXELDATA, "VOXEL_DATA", ICON_TEXTURE, "Voxel Data", ""},
-	{TEX_WOOD, "WOOD", ICON_TEXTURE, "Wood", ""},
+	/* {TEX_POINTDENSITY, "POINT_DENSITY", ICON_TEXTURE, "Point Density", ""}, */
+	/* {TEX_STUCCI, "STUCCI", ICON_TEXTURE, "Stucci", ""}, */
+	/* {TEX_VORONOI, "VORONOI", ICON_TEXTURE, "Voronoi", ""}, */
+	/* {TEX_VOXELDATA, "VOXEL_DATA", ICON_TEXTURE, "Voxel Data", ""}, */
+	/* {TEX_WOOD, "WOOD", ICON_TEXTURE, "Wood", ""}, */
+	{TEX_VRAY, "VRAY", ICON_TEXTURE, "V-Ray", ""},
 	{0, NULL, 0, NULL, NULL}};
 
 #ifdef RNA_RUNTIME

--- source/blender/editors/space_api/CMakeLists.txt	2011-01-06 14:51:57.000000000 +0300
+++ source/blender/editors/space_api/CMakeLists.txt	2011-01-14 18:03:57.457981234 +0300
@@ -23,6 +23,7 @@
 	../include
 	../../blenkernel
 	../../blenlib
+	../../exporter
 	../../makesdna
 	../../makesrna
 	../../windowmanager

--- source/blender/CMakeLists.txt	2011-01-06 14:52:13.000000000 +0300
+++ source/blender/CMakeLists.txt	2011-01-14 17:53:03.215981234 +0300
@@ -41,6 +41,7 @@
 add_subdirectory(modifiers)
 add_subdirectory(makesdna)
 add_subdirectory(makesrna)
+add_subdirectory(exporter)
 
 if(WITH_IMAGE_OPENEXR)
 	add_subdirectory(imbuf/intern/openexr)

--- source/creator/CMakeLists.txt	2011-01-06 14:48:46.000000000 +0300
+++ source/creator/CMakeLists.txt	2011-01-14 19:05:58.851581234 +0300
@@ -491,6 +491,7 @@
		bf_editor_space_time
		bf_editor_space_userpref
		bf_editor_space_view3d
+		vb_mesh_exporter
 
		bf_editor_text
		bf_editor_transform

--- source/blender/editors/render/render_shading.c	2011-01-14 19:55:00.236281234 +0300
+++ source/blender/editors/render/render_shading.c	2011-01-21 18:38:34.927359002 +0300
@@ -579,8 +579,10 @@
 	/* add or copy texture */
 	if(tex)
 		tex= copy_texture(tex);
-	else
+	else {
 		tex= add_texture("Texture");
+		tex->type= TEX_IMAGE;
+	}
 
 	/* hook into UI */
 	uiIDContextProperty(C, &ptr, &prop);

--- source/blender/makesdna/DNA_texture_types.h	2011-01-07 22:42:51.812373001 +0300
+++ source/blender/makesdna/DNA_texture_types.h	2011-01-21 18:42:30.430959002 +0300
@@ -298,6 +298,7 @@
 #define TEX_DISTNOISE	13
 #define TEX_POINTDENSITY	14
 #define TEX_VOXELDATA		15
+#define TEX_VRAY		200
 
 /* musgrave stype */
 #define TEX_MFRACTAL		0
